src/Haskell/Ide/Engine/Server.hs

```diff
@@ -247,9 +256,13 @@ run scheduler _origDir plugins captureFp = flip E.catches handlers $ do
 
 -- ---------------------------------------------------------------------
 
-type ReactorInput
-  = FromClientMessage
-      -- ^ injected into the reactor input by each of the individual callback handlers
+data ReactorInput
+  = CM FromClientMessage
+      -- ^ injected into the reactor input by each of the individual
+      -- callback handlers
+    | PD J.NormalizedUri J.TextDocumentVersion DiagnosticsBySource
+      -- ^ injected into the reactor input by any scheduler needing to
+      -- publish additional diagnostics
```

Comment:
This seems kind of hairy. My original thought that was the "reacted" to messages coming from the client, but this also now handles messages going in both directions?

I agree that there's not really any nice solution to this, but seeing as returning a PD is just calling `liftIO Core.publishDiagnosticsFunc` at the end of the day albeit queued on the reactor thread, would it just be simpler to not bother with `passPublishDiagnostics`? I don't think there is any particular benefit to keeping it on the one thread

Reply:
@bubba the problem is we need to honour the config of max diagnostics to send, which can be changed at any time.

And I would rather do it this way, than have another (global var) IORef in the mix.

Reply:
I don’t think we would need an IORef, just pass a function over to it which is what it seems to be doing already. 
The max diagnostics can also be fetched via the LspFuncs, so I think it should be possible if we just passed that instead

Reply:
Hmm, just took a look at `LspFuncs` again. It has grown quite a bit. And I agree, it has a way to access the config.

And as an aside, given the maxToSend is a config options, maybe we should just automatically use it in the `publishDiagnostics` LspFuncs.

Reply:
Are the LspFuncs updates though in every thread? I could not tell what would update them over the different threads.

Reply:
I think the architecture has mutated to a model where the LspFuncs are basically a lock on the key data structures.  So it should be fine.

I also have a feeling the architecture may change as a result of the ghcide/hie-core collaboration.